ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```
@@ -0,0 +1,592 @@
+            if (type == typeof(KVObject))
+            {
+                ((KVObject)value).Serialize(stringBuilder, indent + 1);
+            }
+            else if (type == typeof(FlaggedString))
+            {
+                stringBuilder.Append((String)value);
+            }
+            else if (type == typeof(String))
+            {
+                String str = (String)value;
+
+                if (str.Contains("\n"))
+                {
+                    stringBuilder.Append("\"\"\"");
+                    stringBuilder.Append(str);
+                    stringBuilder.Append("\"\"\"");
+                }
+                else
+                {
+                    stringBuilder.Append("\"");
+                    stringBuilder.Append(str);
+                    stringBuilder.Append("\"");
+                }
+            }
+            else if (type == typeof(Boolean))
+            {
+                if ((Boolean)value)
+                {
+                    stringBuilder.Append("true");
+                }
+                else
+                {
+                    stringBuilder.Append("false");
+                }
+            }
+            else if (type == typeof(Double))
+            {
+                stringBuilder.Append(((Double)value).ToString("#0.000000",CultureInfo.InvariantCulture));
+            }
+            else if (type == typeof(Int32))
+            {
+                stringBuilder.Append((Int32)value);
+            }
+            else
+            {
+                //This shouldn't happen
+                stringBuilder.Append(value);
+            }
+        }
+    }
+
+    //KV Array extending the KVObject
+    public class KVArray : KVObject
+    {
+        new public List<KVValue> Properties;
+
+        public KVArray(String name) : base(name)
+        {
+            Properties = new List<KVValue>();
+        }
+
+        public override void AddProperty(String name, KVValue value)
```

`name` is silently discarded here. Should `KVArray` really be a subclass of `KVObject` (which acts more like `KVDictionary`)?